File: 04. Fall 2024/CS 4348 - Operating Systems Concepts/Project 01/Driver.java
import java.io.*;
import java.util.ArrayList;
import java.util.List;

public class Driver {
    
    private static List<String> history = new ArrayList<>();

    public static void main(String[] args) {
        if (args.length != 1) {
            System.out.println("Usage: java DriverProgram <logFileName>");
            return;
        }

        String logFileName = args[0];
        try {
            Process loggerProcess = new ProcessBuilder("java", "Logger", logFileName).start();
            Process encryptionProcess = new ProcessBuilder("java", "EncryptionProgram").start();

            BufferedWriter loggerWriter = new BufferedWriter(new OutputStreamWriter(loggerProcess.getOutputStream()));
            BufferedWriter encryptionWriter = new BufferedWriter(new OutputStreamWriter(encryptionProcess.getOutputStream()));
            BufferedReader encryptionReader = new BufferedReader(new InputStreamReader(encryptionProcess.getInputStream()));

            BufferedReader userInput = new BufferedReader(new InputStreamReader(System.in));
            String command;
            while (true) {
                System.out.println("Enter command (password, encrypt, decrypt, history, quit): ");
                command = userInput.readLine().trim().toLowerCase();

                if ("quit".equals(command)) {
                    sendToProcess(loggerWriter, "QUIT");
                    sendToProcess(encryptionWriter, "QUIT");
                    break;
                } else if ("password".equals(command)) {
                    System.out.println("Enter passkey:");
                    String passkey = userInput.readLine();
                    sendToProcess(encryptionWriter, "PASSKEY " + passkey);
                } else if ("encrypt".equals(command)) {
                    System.out.println("Enter text to encrypt:");
                    String text = userInput.readLine();
                    history.add(text);
                    sendToProcess(encryptionWriter, "ENCRYPT " + text);
                    System.out.println(readFromProcess(encryptionReader));
                } else if ("decrypt".equals(command)) {
                    System.out.println("Enter text to decrypt:");
                    String text = userInput.readLine();
                    history.add(text);
                    sendToProcess(encryptionWriter, "DECRYPT " + text);
                    System.out.println(readFromProcess(encryptionReader));
                } else if ("history".equals(command)) {
                    showHistory();
                } else {
                    System.out.println("Unknown command");
                }
            }

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static void sendToProcess(BufferedWriter writer, String message) throws IOException {
        writer.write(message + "\n");
        writer.flush();
    }

    private static String readFromProcess(BufferedReader reader) throws IOException {
        return reader.readLine();
    }

    private static void showHistory() {
        System.out.println("History:");
        for (int i = 0; i < history.size(); i++) {
            System.out.printf("%d: %s%n", i + 1, history.get(i));
        }
    }
}

File: 04. Fall 2024/CS 4348 - Operating Systems Concepts/Project 01/driver.py
# driver.py
import sys
import os
from subprocess import Popen, PIPE

def main(log_file):
    history = []
    logger = Popen(['python', 'logger.py', log_file], stdin=PIPE, encoding='utf8')
    encryption = Popen(['python', 'encryption.py'], stdin=PIPE, stdout=PIPE, encoding='utf8')

    def log(action, message):
        logger.stdin.write(f"{action} {message}\n")
        logger.stdin.flush()
    
    log("START", "Driver started.")
    while True:
        print("Menu:")
        print("1. Set Password")
        print("2. Encrypt")
        print("3. Decrypt")
        print("4. Show History")
        print("5. Quit")
        choice = input("Enter choice: ").strip()

        if choice == "1":  # Set Password
            password = input("Enter password: ").strip()
            encryption.stdin.write(f"PASSKEY {password}\n")
            encryption.stdin.flush()
            result = encryption.stdout.readline().strip()
            log("PASSKEY", result)

        elif choice == "2":  # Encrypt
            text = input("Enter text to encrypt: ").strip()
            history.append(text)
            encryption.stdin.write(f"ENCRYPT {text}\n")
            encryption.stdin.flush()
            result = encryption.stdout.readline().strip()
            print(result)
            log("ENCRYPT", result)
            if result.startswith("RESULT"):
                history.append(result.split(" ", 1)[1])

        elif choice == "3":  # Decrypt
            text = input("Enter text to decrypt: ").strip()
            history.append(text)
            encryption.stdin.write(f"DECRYPT {text}\n")
            encryption.stdin.flush()
            result = encryption.stdout.readline().strip()
            print(result)
            log("DECRYPT", result)
            if result.startswith("RESULT"):
                history.append(result.split(" ", 1)[1])

        elif choice == "4":  # Show History
            print("History:")
            for index, entry in enumerate(history, 1):
                print(f"{index}: {entry}")
            log("HISTORY", "Displayed history")

        elif choice == "5":  # Quit
            encryption.stdin.write("QUIT\n")
            encryption.stdin.flush()
            logger.stdin.write("QUIT\n")
            logger.stdin.flush()
            log("EXIT", "Driver exited.")
            break

if __name__ == "__main__":
    log_file = sys.argv[1]
    main(log_file)
